File: Data Structures & Algorithms/same-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    areSame = True
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if p is None and not q is None:
            self.areSame = False
            return self.areSame
        elif q is None and not p is None:
            self.areSame =False
            return self.areSame

        if p is None and q is None:
            return True

        if not p is None and not q is None:
            if p.val != q.val:
                self.areSame = False
            self.isSameTree(p.left, q.left)
            self.isSameTree(p.right, q.right)
        return self.areSame

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int kthSmallest(TreeNode root, int k) {
        Stack<TreeNode> stack = new Stack<>();
        List<Integer> arr = new ArrayList<>();
        Set<Integer> visited = new HashSet<>();
        while(root != null){
            stack.push(root);
            root = root.left;
        }

        while(!stack.isEmpty()){
            TreeNode currNode = stack.pop();
            arr.add(currNode.val);
            if(!visited.contains(currNode.val)) visited.add(currNode.val);
            if(currNode.right != null && !visited.contains(currNode.right.val)){
                stack.push(currNode.right);
            }
            if(currNode.left != null && !visited.contains(currNode.left.val)){
                stack.push(currNode.left);
            }

            if(arr.size() == k)break;
        }

        return arr.size() == 0 ? 0 : arr.get(k-1); 
    }

}
